app/main.js

```diff
@@ -112,52 +185,62 @@ app.on('activate', () => {
-})
-
+});
 
-function shutdownDaemon(evenIfNotStartedByApp = false) {
-  if (subpy) {
+// When a quit is attempted, this is called, it attempts to shutdown the daemon,
+// and then calls app.quit() to quit for real.
+function shutdownDaemonAndQuit(shutdownEvenIfNotStartedByApp = false) {
+  if (daemonSubprocess) {
     console.log('Killing lbrynet-daemon process');
-    kill(subpy.pid, undefined, (err) => {
+    kill(daemonSubprocess.pid, undefined, (err) => {
       console.log('Killed lbrynet-daemon process');
+      requestedDaemonSubprocessKilled = true;
+      quitNow();
     });
   } else if (evenIfNotStartedByApp) {
-    console.log('Killing lbrynet-daemon, even though app did not start it');
-    client.request('daemon_stop', []);
-    // TODO: If the daemon errors or times out when we make this request, find
-    // the process and force quit it.
+    console.log('Stopping lbrynet-daemon, even though app did not start it');
+    client.request('daemon_stop', [], (err, res) => {
+      if (err) {
+        // We could get an error because the daemon is already stopped (good)
+        // or because it's running but not responding properly (bad).
+        // So try to force kill any daemons that are still running.
+
+        console.log('received error when stopping lbrynet-daemon. Error message: {err.message}');
+        forceKillAllDaemonsAndQuit();
+      } else {
+        console.log('Successfully stopped daemon via RPC call.')
+        quitNow();
+      }
+    });
   } else {
-    console.log('Not killing lbrynet-daemon because app did not start it')
+    console.log('Not killing lbrynet-daemon because app did not start it');
+    quitNow();
   }
 
   // Is it safe to start the installer before the daemon finishes running?
   // If not, we should wait until the daemon is closed before we start the install.
 }
 
-function shutdown() {
-  if (win) {
-    win.loadURL(`file://${__dirname}/dist/quit.html`);
-  }
-  quitting = true;
-  shutdownDaemon();
-}
-
 function upgrade(event, installerPath) {
+  console.log('top of upgrade()')
+
   app.on('quit', () => {
-    shell.openItem(installerPath);
+    console.log('Launching upgrade installer at', installerPath);
+    // This gets triggered called after *all* other quit-related events, so
+    // we'll only get here if we're fully prepared and quitting for real.
+    openItem(installerPath);
   });
+
   if (win) {
     win.loadURL(`file://${__dirname}/dist/upgrade.html`);
   }
-  quitting = true;
-  shutdownDaemon(true);
+
+  app.quit();
   // wait for daemon to shut down before upgrading
   // what to do if no shutdown in a long time?
-  console.log('Update downloaded to ', installerPath);
+  console.log('Update downloaded to', installerPath);
   console.log('The app will close, and you will be prompted to install the latest version of LBRY.');
   console.log('After the install is complete, please reopen the app.');
 }
 
 ipcMain.on('upgrade', upgrade);
-
-ipcMain.on('shutdown', shutdown);
```

Comment:
Nope. The UI used to request the quit after an upgrade, but it was moved into upgrade() here in the main process.